waku/v2/node/storage/message/waku_message_store_queries.nim

```diff
@@ -0,0 +1,365 @@
+
+  if cursor.isSome():  # cursor = senderTimestamp, id, pubsubTopic
+    let senderTimestamp = cursor.get().senderTime
+    checkErr bindParam(s, paramIndex, senderTimestamp)
+    paramIndex += 1
+
+    let id = @(cursor.get().digest.data)
+    checkErr bindParam(s, paramIndex, id)
+    paramIndex += 1
+
+    let pubsubTopic = toBytes(cursor.get().pubsubTopic)
+    checkErr bindParam(s, paramIndex, pubsubTopic)
+    paramIndex += 1
+
+  if pubsubTopic.isSome():
+    let pubsubTopic = toBytes(pubsubTopic.get())
+    checkErr bindParam(s, paramIndex, pubsubTopic)
+    paramIndex += 1
+
+  if startTime.isSome():
+    let time = startTime.get()
+    checkErr bindParam(s, paramIndex, time)
+    paramIndex += 1
+
+  if endTime.isSome():
+    let time = endTime.get()
+    checkErr bindParam(s, paramIndex, time)
+    paramIndex += 1
+
+
+  try:
+    while true:
+      let v = sqlite3_step(s)
+      case v
+      of SQLITE_ROW:
+        onRowCallback(s)
+      of SQLITE_DONE:
+        return ok()
+      else:
+        return err($sqlite3_errstr(v))
+  finally:
+    # release implicit transaction
+    discard sqlite3_reset(s) # same return information as step
+    discard sqlite3_clear_bindings(s) # no errors possible
+
+proc selectMessagesByHistoryQueryWithLimit*(db: SqliteDatabase, 
+                                            contentTopic: Option[seq[ContentTopic]], 
+                                            pubsubTopic: Option[string],
+                                            cursor: Option[Index],  
+                                            startTime: Option[Timestamp],
+                                            endTime: Option[Timestamp],
+                                            limit: uint64, 
+                                            ascending: bool): DatabaseResult[seq[(WakuMessage, Timestamp, string)]] =
+  
+   
+  var messages: seq[(WakuMessage, Timestamp, string)] = @[]
+  proc queryRowCallback(s: ptr sqlite3_stmt) =
+    let 
+      receiverTimestamp = queryRowReceiverTimestampCallback(s, receiverTimestampCol=0)
+      message = queryRowWakuMessageCallback(s, contentTopicCol=1, payloadCol=2, versionCol=4, senderTimestampCol=5)
+      pubsubTopic = queryRowPubsubTopicCallback(s, pubsubTopicCol=3)
+
+    messages.add((message, receiverTimestamp, pubsubTopic))
+
+  let query = block:
+    let
+      contentTopicClause = contentTopicWhereClause(contentTopic)
+      cursorClause = cursorWhereClause(cursor, ascending)
+      pubsubClause = pubsubWhereClause(pubsubTopic)
+      timeRangeClause = timeRangeWhereClause(startTime, endTime)
+    let where = whereClause(contentTopicClause, cursorClause, pubsubClause, timeRangeClause)
+    selectMessagesWithLimitQuery(DbTable, where, limit, ascending)
+
+  let dbStmt = ?db.prepareSelectMessagesWithlimitStmt(query)
+  ?dbStmt.execSelectMessagesWithLimitStmt(
+    contentTopic,
+    pubsubTopic,
+    cursor,
+    startTime,
+    endTime,
+    queryRowCallback
+  )
+
+  ok(messages)
```

Comment:
Here is a bit of a subtle thing from the spec which is not always immediately clear: the `PagingDirection` and the sort order for the messages in the query result are not necessarily the same. In fact, the sorting order _within_ a page must always be _ascending_ according to [the spec](https://rfc.vac.dev/spec/13/).

> "the retrieved page contains an ordered list of WakuMessages from the oldest message to the most recent one."

This is also why the `StoreQueue` has this silly `reverse()` when paging backwards:
https://github.com/status-im/nwaku/blob/fd5da9e42f8a2fcf3b216540657dd947bbb7c466/waku/v2/node/storage/message/waku_store_queue.nim#L234

It causes a bit of inconvenience, but makes sense from a spec point of view as the default. It's up to applications then to modify this sort order to their own needs, but the spec "contract" is that all pages will be sorted ascending. Anyway, long story short, if the paging direction here is `BACKWARD` (i.e. `ascending == false`) I think the messages need to be sorted once more to be ASC in the returned list. For `FORWARD` pages it will already be in the correct order.

Reply:
Sidenote, I think that this logic in the current sqlite-only store has this bug and is why people report reverse chronological messages when connecting to the test fleet using `chat2`.

Reply:
I added the `reverse()` statement when ascending is set to `false`. And updated the test cases accordingly. Now, the issue should be solved